xdp-app-info-snap: Use the desktop ID as app ID and snap ID for permissions

3v1n0 opened the pull request:
From the main commit:

When using snap applications the permissions are saved using a common ID
that is not matching the desktop ID, since they are per-snap, not per
application.

While that is fine,  we were using the very same ID also as the
app ID, which is instead expected to be matching the desktop-ID (without
extension).

To prevent this to happen, use the desktop ID as the app ID and the snap
ID for permission purposes only.

As per this, we can now use the snap desktop ID as the handle that we
expose everywhere, since that's what all the components (shells, and
portal implementations) are expecting to receive.

This fixes an issue existing for long time, that was causing some
portals not to work at all (like the access one) or to misbehave (like
the notification one), while preserving the permission behavior that we
used to have.

Closes: #769

3v1n0 committed:
xdp-app-info: Add separated permissions ID function

While for flatpaks the app-ID matches the value used for permissions
store, in the snap world a snap may contain multiple applications and
they are not fully confined in between each other (processes are, but
they share common data paths), reason why instead of using per-app
permissions we use a per-snap permission.

The snap ID used for permissions is not matching any desktop file or
desktop ID, so this breaks the assumptions of the rest of the stack.

As per this, split the application ID with the permissions ID, by adding
another function that we will re-implement in next commits to have snap
specific behavior.

See: #155 and #769

3v1n0 committed:
xdp-app-info-snap: Use the desktop ID as app ID and snap ID for permissions

When using snap applications the permissions are saved using a common ID
that is not matching the desktop ID, since they are per-snap, not per
application.

While that is fine,  we were using the very same ID also as the
app ID, which is instead expected to be matching the desktop-ID (without
extension).

To prevent this to happen, use the desktop ID as the app ID and the snap
ID for permission purposes only.

As per this, we can now use the snap desktop ID as the handle that we
expose everywhere, since that's what all the components (shells, and
portal implementations) are expecting to receive.

This fixes an issue existing for long time, that was causing some
portals not to work at all (like the access one) or to misbehave (like
the notification one), while preserving the permission behavior that we
used to have.

Closes: #769

3v1n0 committed:
xdp-app-info: Simplify GAppInfo creation

Since we're all in the same page now, we can just create the GAppInfo
using the app ID as base

3v1n0 committed:
xdp-app-info-snap: Remove empty new line

3v1n0 committed:
xdp-app-info-snap: Do not repeat the metadata parsing logic

Use the same code both for tests and real world, so that we can be
closer to reality

3v1n0 commented:
/cc @aleasto @jhenstridge

3v1n0 commented on src/xdp-app-info.h:
```diff
@@ -47,6 +47,8 @@ gboolean xdp_app_info_is_host (XdpAppInfo *app_info);
 
 const char * xdp_app_info_get_id (XdpAppInfo *app_info);
 
+const char * xdp_app_info_get_permissions_id (XdpAppInfo *app_info);
```
Well, reason why of commit e225c34f6a851be9be99409c03fd1c0b3fef7934 and thus the usage of the API is very limited to that scope or to debug places, but I also was unsure how to name it...

Or we could make the API even more "private", and only include it inside the permission code.

3v1n0 commented:
> I think that would be a good idea - perhaps `xdp_app_info_get_permissions_id` or even `xdp_app_info_get_trust_domain` for the trust domain (snap app ID or Flatpak app ID), and `xdp_app_info_get_desktop_id` for the one that is specifically the .desktop file? That way, anyone calling one of these functions is forced to acknowledge that there is a question about which one they want, and answer it by choosing the appropriate one.

Indeed that is a good proposal, @GeorgesStavracas (or @swick) maybe you can comment here and I could move the refactoring commits to another PR where we just move things around, and then here we only handle the specific snap cases

smcv commented:
> especially if one of them is "the app" and the other is a sandboxable thumbnailer or something

For comparison, in Flatpak-world, this would be implemented by having the user-facing part of the app talk to portals in the usual way, but preventing the sandboxable thumbnailer from talking to portals at all (`flatpak-spawn --sandbox` doesn't give D-Bus access, AIUI).

swick commented on src/wallpaper.c:
```diff
@@ -212,7 +204,7 @@ handle_set_wallpaper_in_thread_func (GTask *task,
            */
           g_assert (xdp_app_info_is_host (request->app_info));
-          app_id = g_strdup ("");
+          app_id = "";
```
isn't app_id and id the same now?

swick commented on tests/xdp_utils.py:
```diff
@@ -382,8 +382,9 @@ def new_snap(
-        app_id = f"snap.{snap_name}"
-        desktop_file = f"{snap_name}_{app_name}.desktop"
+        permissions_id = f"snap.{snap_name}"
+        app_id = f"{snap_name}_{app_name}.desktop"
```
is this a typo or is the app_id really supposed to include the .desktop suffix?

swick reviewed (commented):
Yay for looking into this! I really like that this fixes up the tests for snap and manages to hide the permission_id detail in `xdp-permissions.h`.

mcatanzaro commented:
> Dunno if this actually helps with anything because it's still just the thing we currently call the app id, even if we now start using another value for storing the permissions.

This problem is already solved by changing the xdp permissions functions to take the entire `XdpAppInfo *` rather than a string, so we don't have to worry about it anymore.

swick commented:
I think to get this merged, what I want is...

* concrete examples in the commit message (what @smcv is asking for)
* an ack from @jhenstridge or someone else from the snap side

jadahl commented:
@3v1n0, how do you imagine places that check the permission store outside of xdg-desktop-portal doing? E.g. https://gitlab.freedesktop.org/pipewire/wireplumber/-/blob/15f5f96693d6155750db3713d696eb41446fadbc/src/scripts/client/access-portal.lua#L60

Should portal backends be discouraged to handle integrating with the permission store themselves?

swick commented:
There are a bunch of commits in this PR which would be nice to have, no matter the decision about what to use in the permission store. Could you split them out into its own PR? I think only commits 46b69f7a112e6e04271b12f1a9679bdbfe5720b0 and 0e1f29089a2f2f1a9007f5f1a74d3c1fa39eb96c change the semantics.

3v1n0 commented:
> concrete examples in the commit message (what @smcv is asking for)

Ok, sure... I will try to add also a test with more complex cases.

Do you have a deadline for the re-review and release?

jadahl commented:
> Well, using desktop ID in these cases should be fine, it's likely not respecting the real permissions but it should good enough (as it's better to show a pointless prompt more, than just not work).

I think for the camera portal (which `access-portal.lua` is currently only about) will just deny access if the app ID added to the permission store in https://github.com/flatpak/xdg-desktop-portal/blob/7af478c5448939c6a99fb5728dd4ff3dee8ecf36/src/camera.c#L150. I see that you're now changing to set it to the "snap ID" however, so I guess it will still work as expected, so as long as wireplumber doesn't assume it can use the "app ID" as anything that can match to a desktop file no one will notice.

3v1n0 commented:
> . I see that you're now changing to set it to the "snap ID" however, so I guess it will still work as expected, so as long as wireplumber doesn't assume it can use the "app ID" as anything that can match to a desktop file no one will notice.

So in terms of permissions it should work, although I noticed that camera portal does not work as expected at the moment, unlike when doing a "simpler" change such as https://github.com/3v1n0/xdg-desktop-portal/commits/snap-initial-fixes/ so let me check it further, because in both branches we're should be sending the same permission ID, while the desktop ID is used by the shell must be different

jadahl commented:
> They might have the expectation that what's stored there is the app id which maps to a desktop file and not potentially another thing like the permission id we want to introduce here.

Right; seems we don't really do much with the permission store in xdp-gnome, gnome-shell and localsearch (which are the three places I checked), but would there be checks that require GUI, whatever is passed should also usable to resolve a desktop file describing the app.

> Similarly, existing permissions for snaps will become useless. Might even have to check if the new permission id can possibly collide with the previous app id and result in some confused permission.

If this changes what is stored in the permission store for snaps, it'd be useful to have a migration script of sorts, like we've had for similar situations (in other projects) when stored formats change.

3v1n0 commented:
@jadahl so https://github.com/flatpak/xdg-desktop-portal/pull/1764/commits/72bad0f56654537b1292c4fb771dae28fbc0f43b is enough here for pipewire to work, but as said is probably better to generalize it for the permissions portal in general

3v1n0 committed:
document-portal: Ignore unused app_id value

We get the permission ID anyways from the bus

3v1n0 committed:
document-store: Get permissions ID to compute the permissions

3v1n0 committed:
document-portal/permission-db: Use permissions ID as argument instead of app

We only change this in the public API, not to clutter the diff with
variables renames, but the permission DB should now handle just apps
permissions IDs

3v1n0 committed:
document-portal: Use app permissions ID instead of app ID for permissions

When getting or setting an application permission, we relied on its app
ID, but as per the introduction of the permissions ID we should now use
it instead.

The value is just really an app ID for flatpaks while it's based on the
snap name for snaps

3v1n0 committed:
document-portal/permissions-db: Drop unused functions

3v1n0 committed:
tests/permissions-store: Use XdpAppInfo permissions ID

Instead of using undefined applications, let's use the one we have
defined for the test so that we can rely on the actual permissions ID

3v1n0 committed:
document-portal/xdg-permission-store: Use app permissions ID

This is a preparatory commit for future changes, but in general we
should use the permission ID, not the app ID everywhere

3v1n0 committed:
tests/documents: Run the tests for all the kinds using app permissions ID

Sodivad commented:
> I think Jonas has a point with others components poking at the permission store. They might have the expectation that what's stored there is the app id which maps to a desktop file and not potentially another thing like the permission id we want to introduce here. That actually needs checking.

> Right; seems we don't really do much with the permission store in xdp-gnome, gnome-shell and localsearch (which are the three places I checked), but would there be checks that require GUI, whatever is passed should also usable to resolve a desktop file describing the app.

gnome settings does however https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/main/panels/applications/cc-applications-panel.c?ref_type=heads#L265

I didn't follow the entire discussion but is that the correct id to use for snaps? I do the same  in Plasma FWIW. It would also good if the we could document which ids will be used for permissions and how they relate to the `app_id` that is passed to the backend. Also for something like https://github.com/flatpak/xdg-desktop-portal/pull/1768 that could require the backend to look into the permission store.

swick commented:
I think we really need to standardize some new fields in desktop files, namely:

* AppID
* CollectionID
* EngineID
* PermissionID

Where The AppID is the desktop file name if not specified (we should still recommend the desktop file to match this; it's just unfortunate that snap doesn't follow this).

The CollectionID would be the source, so org.flathub for example, or io.snapcraft for snaps.

The EngineID would be org.flatpak or io.snapcraft, or nothing for host apps.

The PermissionID would be the same as the AppID if not specified.

But even if we do that, we have to find a way to deal with previous versions of desktop files.

3v1n0 committed:
xdg-permission-store: Always use the permission ID to get or set Apps permissions

Snaps use a permssion ID that does not match the application ID, while
the permissions store portal API refers to the app (IDs).

Now, we do not want to change this, given that portals assume that what
they receive the app-id, that is the basename of the desktop ID of the
application, so we do it here.

When an app request is received, we check the desktop file of the given
app and if that belongs to a snap application, then we use the snap
name as the permission id.

In this way we can both support the standard flatpak apps and the snap
ones, without changing the way they saved the permissions.

This implies some minimal overhead for non-snap applications, since we
are now looking up into desktop files, but it should be quite irrelevant

Potentially we could, in future, just track the files in the well known
paths in which the sandboxed applications put their desktop files, so
that we can optimize this a bit with an hashmap.

3v1n0 committed:
document-portal: Do not try to use alternate method for file access for snaps

It's all been designed around flatpak internals, so it won't ever work
with snaps.

Just return early if snap routine file-access fails

3v1n0 committed:
document-portal: Do not check the file-access for full permission IDs

In case a full permission ID (snap + snap app) is used, then snap
routine would fail, so prevent this by just getting the snap name

3v1n0 committed:
document-portal: Actually make use of an error on spawn

We allocate an error for spawning, but we do not use it at all.

Let's add some debug messages at least

3v1n0 committed:
xdp-utils: Avoid complex checks on snap permission IDs

3v1n0 committed:
document-portal: Convert early from app-id to permissions-id

In this way we avoid doing it at the end of the chain in the permission
portal

3v1n0 committed:
xdp-permissions: Use permissions ID to save and check the permissions

While this just the same at the moment, we are going to split the IDs
for the snap case, so we need to ensure that everywhere a permission is
involved we always refer to the permission ID

3v1n0 committed:
tests/documents: Test the usage of permission IDs and app IDs

We're going to make possible to support both the permissions ID and App ID
as handle to change an application permission, so we need to ensure that
both works as expected.

Thus add tests that use both IDs

3v1n0 committed:
snap: Use per-sub-app permissions ID for better granularity

When a snap sub-app has a different app name than the snap instance
name, use "snap.<instance>_<appname>" as the permissions ID instead of
the plain "snap.<instance>".

For the main cases (where the app name matches the snap instance
name), the existing "snap.<instance>" ID is preserved, keeping full
backward compatibility and avoiding permission churn for existing users.

Sub-apps that previously fell back to snap-level permissions will no
longer share them, which is intentional: each sub-app should own its
permissions independently.

The logic is applied consistently in both the snap metadata parser
and the desktop-file based permissions-ID lookup.

Although the files permissions are still shared, because of snapd
limitation.

A new SNAP_SUB_APP test fixture to verify the sub-app code path
explicitly.

See: https://github.com/flatpak/xdg-desktop-portal/pull/798

3v1n0 committed:
camera: Use permission ID to request pipewire to open the stream

Pipewire uses the permission portal to check if the application can
access by looking up the permission IDs, so we need to pass the
permission ID to it

3v1n0 commented on document-portal/document-portal.c:
```diff
@@ -647,8 +647,13 @@ app_has_file_access (const char *target_permissions_id,
+
+      parts = g_strsplit (target_permissions_id + strlen ("snap."), "_", -1);
+
+      /* Snap does not support permission check for sub-applications */
```
Indeed, but `snap routine` check does not do that, I feel we should avoid using it though, and just use the REST api here...

@swick how would you feel about having libsoup and json-glib as dependencies (or optional ones)? :)

Or we should improve the tool in snapd instead.

swick replied:
Not entirely opposed, especially fine with a JSON dependency. What would you need libsoup for though?

aleasto replied:
for talking to the snapd REST API instead of shelling-out to `snap routine portal-info`

3v1n0 replied:
Yeah, basically snapd has a [REST API](https://snapcraft.io/docs/reference/development/snapd-rest-api/) that we can use through an UNIX socket.

This could be done via [snapd-glib](https://github.com/canonical/snapd-glib/) or more manually using libsoup directly and some JSON lib.

swick replied:
Alright, fine with me.

3v1n0 replied:
Ok, I'll do that in another PR though, since it would imply adjusting tests, so I'd prefer not to make this even more complex.

aleasto replied:
We found out by talking to the snapd people that while permissions are declared on a per-app basis, they're only actually granted/revoked on a per-snap basis. So this already matches expectations.

swick commented on document-portal/permission-db.c:
```diff
@@ -49,11 +49,11 @@ struct PermissionDb
   gboolean   dirty;
 
-  /* Map id => GVariant (data, sorted-dict[appid->perms]) */
+  /* Map id => GVariant (data, sorted-dict[app_permissions_id->perms]) */
```
Can you please also check the docs to make sure we don't describe an outdated model?

swick commented on src/camera.c:
```diff
@@ -225,15 +225,15 @@ handle_access_camera (XdpDbusCamera *object,
 
   pipewire_properties =
-    pw_properties_new ("pipewire.access.portal.app_id", app_id,
+    pw_properties_new ("pipewire.access.portal.app_id", permissions_id,
```
There is a PR which moves the pipewire plugin into the repo so we could rename this appropriately.

swick commented on document-portal/permission-db.c:
```diff
@@ -1248,3 +1249,40 @@ permission_db_entry_print_string (PermissionDbEntry *entry,
 }
+
+char *
+permissions_id_for_app_id (const char *app_id)
```
I think this is the big pain point for me in this PR. I don't like that this isn't abstracted away in the XdpAppInfo but then again the XdpAppInfo is about a specific instance. Fundamentally the problem here seems to be that we get passed app ids when we would need permissions ids.

I'm willing to accept this code here, *if* it is documented clearly that this is a stop-gap measure, and you sign up to do the work to convert the callers to provide permissions ids.

swick reviewed (commented):
This also needs a rebase because I changed the pytest XdpAppInfo again.

swick commented:
The changes are a bit much to still get them into the 1.22 stable release.

aleasto commented on src/xdp-app-info-snap.c:
```diff
@@ -255,7 +257,19 @@ parse_snap_metadata (GKeyFile         *metadata,
+    }
+  else
+    {
+      permissions_id = g_strconcat ("snap.", snap_name, "_", snap_app_name, NULL);
```
Ok, to avoid having to migrate existing permission stores, let's keep using `_`; it's still fine, as to parse back into snap components we can simply split on the last `_` character instead of the first.

aleasto commented on document-portal/xdg-permission-store.c:
```diff
@@ -397,12 +401,16 @@ handle_set (XdgPermissionStore     *object,
 
       old_entry = new_entry;
-      new_entry = permission_db_entry_set_app_permissions (new_entry, child_app_id, (const char **) permissions);
+      child_permissions_id = child_app_id;
```
why are you not translating `app_id` -> `permissions_id` here?

aleasto reviewed (changes requested):
if xdg-permission-store implicitly translates app ids to permissions ids, then users calling `Lookup()` will be unable to index the returned map by what they thought would be app ids.

that might be fine to break the interface like this, as the only user shall be x-d-p itself, but then we need to adjust all the callers of `org.freedesktop.impl.portal.PermissionStore.Lookup`

This is causing bug https://bugs.launchpad.net/ubuntu/+source/xdg-desktop-portal/+bug/2152906

aleasto commented:
> that might be fine to break the interface like this, as the only user shall be x-d-p itself

Actually that's not true. It is also used to implement settings in apps like gnome-control-center

aleasto commented:
> > that might be fine to break the interface like this, as the only user shall be x-d-p itself
> 
> Actually that's not true. It is also used to implement settings in apps like gnome-control-center

I suppose that's why you have https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/3353
Maybe you can argue this never worked previously for snaps, so it is not a breaking change.

Sodivad commented:
I have the equivalent code as gnome-settings in plasma as well

swick commented:
@aleasto I'm a bit lost on the state of this PR. Is this done and you're just waiting for a review or are you still trying to iron out some edge cases?

aleasto commented:
This needs addressing my earlier comment https://github.com/flatpak/xdg-desktop-portal/pull/1764#pullrequestreview-4561265649 at least, because it's a regression.

I believe @3v1n0 will address that sometime soon.
